cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/MaintenanceEventDetector.java

```diff
@@ -6,42 +6,112 @@
   public static final long DETECTION_NOT_READY_BACKOFF_MS = TimeUnit.SECONDS.toMillis(10);
+  public static final long IDEMPOTENCE_CACHE_CLEANER_PERIOD_SECONDS = 5;
+  public static final long IDEMPOTENCE_CACHE_CLEANER_INITIAL_DELAY_SECONDS = 0;
   // TODO: Make this configurable.
   public static final Duration READ_EVENTS_TIMEOUT = Duration.ofSeconds(5);
   private volatile boolean _shutdown;
   private final MaintenanceEventReader _maintenanceEventReader;
+  private final boolean _idempotenceEnabled;
+  private final Duration _idempotenceRetention;
+  private final ScheduledExecutorService _idempotenceCacheCleaner;
+  private final Map<Long, MaintenanceEvent> _maintenanceEventByTime;
 
   public MaintenanceEventDetector(Queue<Anomaly> anomalies, KafkaCruiseControl kafkaCruiseControl) {
     super(anomalies, kafkaCruiseControl);
     KafkaCruiseControlConfig config = _kafkaCruiseControl.config();
     _shutdown = false;
     Map<String, Object> configWithCruiseControlObject = Collections.singletonMap(KAFKA_CRUISE_CONTROL_OBJECT_CONFIG,
                                                                                  kafkaCruiseControl);
-    _maintenanceEventReader = config.getConfiguredInstance(MAINTENANCE_EVENT_READER_CLASS_CONFIG,
+    _maintenanceEventReader = config.getConfiguredInstance(AnomalyDetectorConfig.MAINTENANCE_EVENT_READER_CLASS_CONFIG,
                                                            MaintenanceEventReader.class,
                                                            configWithCruiseControlObject);
+    _idempotenceEnabled = config.getBoolean(AnomalyDetectorConfig.MAINTENANCE_EVENT_ENABLE_IDEMPOTENCE_CONFIG);
+    _idempotenceRetention = Duration.ofMillis(config.getLong(AnomalyDetectorConfig.MAINTENANCE_EVENT_IDEMPOTENCE_RETENTION_MS_CONFIG));
+    if (_idempotenceEnabled) {
+      int maxIdempotenceCacheSize = config.getInt(AnomalyDetectorConfig.MAINTENANCE_EVENT_MAX_IDEMPOTENCE_CACHE_SIZE_CONFIG);
+      _maintenanceEventByTime = new LinkedHashMap<Long, MaintenanceEvent>() {
+        @Override
+        protected boolean removeEldestEntry(Map.Entry<Long, MaintenanceEvent> eldest) {
+          return this.size() > maxIdempotenceCacheSize;
+        }
+      };
+      _idempotenceCacheCleaner = Executors.newSingleThreadScheduledExecutor(new KafkaCruiseControlThreadFactory("IdempotenceCacheCleaner"));
+      _idempotenceCacheCleaner.scheduleAtFixedRate(new IdempotenceCacheCleaner(),
+                                                   IDEMPOTENCE_CACHE_CLEANER_INITIAL_DELAY_SECONDS,
+                                                   IDEMPOTENCE_CACHE_CLEANER_PERIOD_SECONDS,
+                                                   TimeUnit.SECONDS);
+    } else {
+      _maintenanceEventByTime = null;
+      _idempotenceCacheCleaner = null;
+    }
+  }
+
+  private synchronized void removeOldEventsFromIdempotenceCache() {
+    LOG.debug("Remove old events from idempotence cache.");
```

Comment:
Maybe we should log something after we remove some entries and include the number of entries we have removed in this method

Reply:
I agree that logging the `idempotence cache` before and after to see if there are any issues with the cache or cleanup mechanism might be useful for debugging purposes.